File: SubCameraApp_refactory_260311/testing_app/src/core/ConfigLoader.cpp
#include "ConfigLoader.h"
#include <fstream>
#include <iostream>

bool ConfigLoader::load(const std::string& path) {
    std::ifstream f(path);
    if (!f.is_open()) {
        std::cerr << "[ConfigLoader] 파일 열기 실패: " << path << "\n";
        return false;
    }
    try {
        f >> json_;
    } catch (const std::exception& e) {
        std::cerr << "[ConfigLoader] JSON 파싱 오류: " << e.what() << "\n";
        return false;
    }

    for (auto it = json_.begin(); it != json_.end(); ++it) {
        if (it.key().front() != '_' && it->is_object()) {
            parseSection(it.key());
        }
    }
    std::cout << "[ConfigLoader] 로드 완료: " << path << "\n";
    return true;
}

void ConfigLoader::parseSection(const std::string& section) {
    if (!json_.contains(section)) return;
    auto& sec = json_[section];
    for (auto it = sec.begin(); it != sec.end(); ++it) {
        ParamEntry e;
        auto& obj = it.value();
        if (obj.contains("value")) {
            if (obj["value"].is_string()) {
                e.strValue = obj["value"].get<std::string>();
            } else {
                e.value = obj["value"].get<double>();
            }
        }
        if (obj.contains("default")) {
            if (obj["default"].is_string()) {
                e.strValue = obj["default"].get<std::string>();
            } else {
                e.def = obj["default"].get<double>();
            }
        }
        if (obj.contains("min"))  e.minVal = obj["min"].get<double>();
        if (obj.contains("max"))  e.maxVal = obj["max"].get<double>();
        if (obj.contains("desc")) e.desc   = obj["desc"].get<std::string>();
        params_[section][it.key()] = e;
    }
}

bool ConfigLoader::save(const std::string& path) const {
    nlohmann::json out = json_;
    // value 필드만 현재 파라미터로 업데이트
    for (auto& [section, keys] : params_) {
        for (auto& [key, entry] : keys) {
            if (out.contains(section) && out[section].contains(key)) {
                if (!entry.strValue.empty()) {
                    out[section][key]["value"] = entry.strValue;
                } else {
                    out[section][key]["value"] = entry.value;
                }
            }
        }
    }
    std::ofstream f(path);
    if (!f.is_open()) {
        std::cerr << "[ConfigLoader] 저장 실패: " << path << "\n";
        return false;
    }
    f << out.dump(2);
    std::cout << "[ConfigLoader] 저장 완료: " << path << "\n";
    return true;
}

double ConfigLoader::get(const std::string& section, const std::string& key) const {
    auto sit = params_.find(section);
    if (sit == params_.end()) return 0.0;
    auto kit = sit->second.find(key);
    if (kit == sit->second.end()) return 0.0;
    return kit->second.value;
}

void ConfigLoader::set(const std::string& section, const std::string& key, double val) {
    params_[section][key].value = val;
}

std::string ConfigLoader::getStr(const std::string& section, const std::string& key) const {
    auto sit = params_.find(section);
    if (sit == params_.end()) return "";
    auto kit = sit->second.find(key);
    if (kit == sit->second.end()) return "";
    return kit->second.strValue;
}

void ConfigLoader::setStr(const std::string& section, const std::string& key, const std::string& val) {
    params_[section][key].strValue = val;
}

const std::unordered_map<std::string, ParamEntry>&
ConfigLoader::getSection(const std::string& section) const {
    static std::unordered_map<std::string, ParamEntry> empty;
    auto it = params_.find(section);
    return (it != params_.end()) ? it->second : empty;
}

double ConfigLoader::getMin(const std::string& s, const std::string& k) const {
    auto sit = params_.find(s);
    if (sit == params_.end()) return 0.0;
    auto kit = sit->second.find(k);
    return (kit != sit->second.end()) ? kit->second.minVal : 0.0;
}

double ConfigLoader::getMax(const std::string& s, const std::string& k) const {
    auto sit = params_.find(s);
    if (sit == params_.end()) return 1.0;
    auto kit = sit->second.find(k);
    return (kit != sit->second.end()) ? kit->second.maxVal : 1.0;
}

double ConfigLoader::getDefault(const std::string& s, const std::string& k) const {
    auto sit = params_.find(s);
    if (sit == params_.end()) return 0.0;
    auto kit = sit->second.find(k);
    return (kit != sit->second.end()) ? kit->second.def : 0.0;
}

File: SubCameraApp_refactory_260311/testing_app_python_v2/core/config_loader.py
"""
config_loader.py – JSON 파라미터 로드/저장 (ConfigLoader Python 포트)
"""
import json
import os
from dataclasses import dataclass, field
from typing import Any, Dict


@dataclass
class ParamEntry:
    value: Any = 0.0
    default: Any = 0.0
    min_val: float = 0.0
    max_val: float = 0.0
    desc: str = ""


class ConfigLoader:
    """testing_params.json 로드/저장 및 section.key 기반 파라미터 접근"""

    def __init__(self):
        self._params: Dict[str, Dict[str, ParamEntry]] = {}
        self._raw: dict = {}
        self._path: str = ""

    # ------------------------------------------------------------------
    def load(self, path: str) -> bool:
        self._path = path
        try:
            with open(path, "r", encoding="utf-8") as f:
                self._raw = json.load(f)
        except Exception as e:
            print(f"[ConfigLoader] 로드 실패: {e}")
            return False

        for section, body in self._raw.items():
            if section.startswith("_") or not isinstance(body, dict):
                continue
            self._params[section] = {}
            for key, obj in body.items():
                if not isinstance(obj, dict):
                    continue
                e = ParamEntry()
                e.value   = obj.get("value",   0.0)
                e.default = obj.get("default", 0.0)
                e.min_val = obj.get("min",     0.0)
                e.max_val = obj.get("max",     1.0)
                e.desc    = obj.get("desc",    "")
                self._params[section][key] = e
        print(f"[ConfigLoader] 로드 완료: {path}")
        return True

    # ------------------------------------------------------------------
    def save(self, path: str | None = None) -> bool:
        target = path or self._path
        out = json.loads(json.dumps(self._raw))  # deep copy
        for section, keys in self._params.items():
            for key, entry in keys.items():
                if section in out and key in out[section]:
                    out[section][key]["value"] = entry.value
        try:
            with open(target, "w", encoding="utf-8") as f:
                json.dump(out, f, indent=2, ensure_ascii=False)
            print(f"[ConfigLoader] 저장 완료: {target}")
            return True
        except Exception as e:
            print(f"[ConfigLoader] 저장 실패: {e}")
            return False

    # ------------------------------------------------------------------
    def get(self, section: str, key: str, fallback: float = 0.0) -> float:
        try:
            return float(self._params[section][key].value)
        except (KeyError, TypeError, ValueError):
            return fallback

    def get_str(self, section: str, key: str, fallback: str = "") -> str:
        try:
            return str(self._params[section][key].value)
        except KeyError:
            return fallback

    def set(self, section: str, key: str, value: Any):
        if section in self._params and key in self._params[section]:
            self._params[section][key].value = value

    def get_min(self, section: str, key: str) -> float:
        try: return self._params[section][key].min_val
        except KeyError: return 0.0

    def get_max(self, section: str, key: str) -> float:
        try: return self._params[section][key].max_val
        except KeyError: return 1.0

    def get_default(self, section: str, key: str) -> Any:
        try: return self._params[section][key].default
        except KeyError: return 0.0

    def get_entry(self, section: str, key: str) -> ParamEntry | None:
        return self._params.get(section, {}).get(key)

    def sections(self):
        return list(self._params.keys())

    def keys(self, section: str):
        return list(self._params.get(section, {}).keys())
